data set 설명

bus: 전국 버스 정류장 데이터 불러옴

bus_jeju: bus에서 제주만 필터링함

jeju_tourist: 제주관광빅데이터플랫폼에서 받아온 많이 방문하는 top30 관광지

jeju_tourist_filtered: 방문횟수, 관광지 열 나눔

jeju_tour_fin: 필요없는 열 제거

jeju_tour_alter: geopy 인식 안되는 지명 바꿈

jeju_tour_loclat.csv: 위도 경도 변환 완료

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

bus = pd.read_csv('/content/drive/MyDrive/SeSAC/국토교통부_전국 버스정류장 위치정보_20241028.csv', encoding='cp949')

전국 버스 정류장 데이터를 제주에 있는 것만 필터링 할 예정

In [5]:
bus.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206022 entries, 0 to 206021
Data columns (total 9 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   정류장번호    206022 non-null  object 
 1   정류장명     206022 non-null  object 
 2   위도       206019 non-null  float64
 3   경도       206019 non-null  float64
 4   정보수집일    206022 non-null  object 
 5   모바일단축번호  188778 non-null  float64
 6   도시코드     206022 non-null  int64  
 7   도시명      206022 non-null  object 
 8   관리도시명    206022 non-null  object 
dtypes: float64(3), int64(1), object(5)
memory usage: 14.1+ MB


In [6]:
bus.head()

정류장번호         정류장명         위도          경도       정보수집일   모바일단축번호  \
0  ADB354000001        길안정류장  36.458658  128.891228  2024-10-28  540001.0   
1  ADB354000002  고란.계명산휴양림입구  36.397708  128.924029  2024-10-28  540002.0   
2  ADB354000003          금곡리  36.397504  128.926821  2024-10-28  540003.0   
3  ADB354000004         금소리1  36.509449  128.849671  2024-10-28  540004.0   
4  ADB354000005         금소리5  36.503629  128.854783  2024-10-28  540005.0   

    도시코드       도시명  관리도시명  
0  37040  경상북도 안동시  안동BIS  
1  37040  경상북도 안동시  안동BIS  
2  37040  경상북도 안동시  안동BIS  
3  37040  경상북도 안동시  안동BIS  
4  37040  경상북도 안동시  안동BIS

info()를 통해서 정보를 보니까 도시명에서 '제주'라는 걸 찾아내서 필터링 하면 될것 같음

하지만 '제주특별자치도'일지, '제주도'일지, '제주특별자치도 제주시, 서귀포시' 이런식일지를 몰라서 '제주'라는 단어를 갖고만 있으면 필터링 할 수 있게함

사실 도시코드로 하면 좋지만 이렇게 되면 206022개의 row 중에 제주를 찾아서 코드를 찾아야되는데 엑셀 파일이 커서 귀찮음

In [7]:
# 도시명 열에서 '제주'라는 단어가 포함된 행만 필터링
bus_jeju = bus[bus['도시명'].str.contains('제주', na=False)]

# 필터링된 결과를 새로운 CSV 파일로 저장
bus_jeju.to_csv('bus_jeju.csv', index=False)

print("필터링된 데이터를 'bus_jeju.csv' 파일로 저장했습니다.")

필터링된 데이터를 'bus_jeju.csv' 파일로 저장했습니다.


전국 버스 정류장 데이터 중에 제주에 있는 것만 필터링 해서 저장


In [8]:
jeju_tourist = pd.read_csv('/content/drive/MyDrive/SeSAC/tourist.csv', encoding = 'utf-16 ')

In [9]:
jeju_tourist.head()

Area Nm\tURL\tCnt
0  협재해수욕장\thttps://www.visitjeju.net/kr/detail/vi...
1  함덕해수욕장\thttps://www.visitjeju.net/kr/detail/vi...
2  한라산영실코스\thttps://www.visitjeju.net/kr/detail/v...
3  카멜리아힐\thttps://www.visitjeju.net/kr/detail/vie...
4  천지연폭포\thttps://www.visitjeju.net/kr/detail/vie...

In [10]:
print(jeju_tourist.columns)

Index(['Area Nm\tURL\tCnt'], dtype='object')


In [11]:
# jeju_tourist 데이터프레임의 'cnt' 열을 문자열로부터 숫자만 추출하여 새로운 열 생성
jeju_tourist['Cnt'] = jeju_tourist['Area Nm\tURL\tCnt'].str.split().str[-1].astype(int)
# jeju_tourist 데이터프레임의 'area' 열을 문자열로부터 첫 번째 항목(관광지 이름) 추출
jeju_tourist['area'] = jeju_tourist['Area Nm\tURL\tCnt'].str.split('\t').str[0]

# jeju_tourist 데이터프레임을 'cnt' 열 기준으로 내림차순 정렬
jeju_tourist_filtered = jeju_tourist.sort_values(by='Cnt', ascending=False)

# 필터링된 결과 출력
print(jeju_tourist_filtered)


                                    Area Nm\tURL\tCnt   Cnt  \
9                                      제주동화마을\t\t7327  7327   
27  비자림\thttps://www.visitjeju.net/kr/detail/view?...  6745   
2   한라산영실코스\thttps://www.visitjeju.net/kr/detail/v...  6381   
26  사려니숲길\thttps://www.visitjeju.net/kr/detail/vie...  6000   
21  성산일출봉(UNESCO 세계자연유산)\thttps://www.visitjeju.ne...  5679   
8   제주민속오일시장\thttps://www.visitjeju.net/kr/detail/...  4417   
13  용머리해안\thttps://www.visitjeju.net/kr/detail/vie...  3852   
12  우도(해양도립공원)\thttps://www.visitjeju.net/kr/detai...  3510   
22  섭지코지\thttps://www.visitjeju.net/kr/detail/view...  3235   
28  만장굴\thttps://www.visitjeju.net/kr/detail/view?...  2889   
29  동문재래시장\thttps://www.visitjeju.net/kr/detail/vi...  2854   
19                                     송악산둘레길\t\t2533  2533   
24  새별오름\thttps://www.visitjeju.net/kr/detail/view...  2521   
5   주상절리대(중문대포해안)\thttps://www.visitjeju.net/kr/de...  2488   
0   협재해수욕장\thttps://www.visitjeju.net/kr/detail/vi...  

In [12]:
jeju_tourist_filtered.head()

Area Nm\tURL\tCnt   Cnt  \
9                                      제주동화마을\t\t7327  7327   
27  비자림\thttps://www.visitjeju.net/kr/detail/view?...  6745   
2   한라산영실코스\thttps://www.visitjeju.net/kr/detail/v...  6381   
26  사려니숲길\thttps://www.visitjeju.net/kr/detail/vie...  6000   
21  성산일출봉(UNESCO 세계자연유산)\thttps://www.visitjeju.ne...  5679   

                    area  
9                 제주동화마을  
27                   비자림  
2                한라산영실코스  
26                 사려니숲길  
21  성산일출봉(UNESCO 세계자연유산)

In [13]:
jeju_tour_fin = jeju_tourist_filtered.drop(['Area Nm\tURL\tCnt'],axis=1)

In [14]:
jeju_tour_fin.head()

Cnt                  area
9   7327                제주동화마을
27  6745                   비자림
2   6381               한라산영실코스
26  6000                 사려니숲길
21  5679  성산일출봉(UNESCO 세계자연유산)

In [15]:
jeju_tour_fin

Cnt                  area
9   7327                제주동화마을
27  6745                   비자림
2   6381               한라산영실코스
26  6000                 사려니숲길
21  5679  성산일출봉(UNESCO 세계자연유산)
8   4417              제주민속오일시장
13  3852                 용머리해안
12  3510            우도(해양도립공원)
22  3235                  섭지코지
28  2889                   만장굴
29  2854                동문재래시장
19  2533                송악산둘레길
24  2521                  새별오름
5   2488         주상절리대(중문대포해안)
0   2345                협재해수욕장
18  2301                   쇠소깍
4   2134                 천지연폭포
1   2112                함덕해수욕장
15  1830             애월한담해안산책로
17  1740             수월봉 지질트레일
10  1738                  정방폭포
3   1613                 카멜리아힐
11  1495            윗세오름(웃세오름)
16  1483              신창풍차해안도로
7   1468               제주아트서커스
14  1440               오설록티뮤지엄
25  1411                  산굼부리
23  1396             서귀포매일올레시장
6   1369             제주절물자연휴양림
20  1366                성이시돌목장

In [16]:
jeju_tour_fin.to_csv('jeju_tour_fin.csv', index=False)

In [18]:
import pandas as pd
jeju_tour_alter = jeju_tour_fin.copy()
# 기존 jeju_tour_fin 데이터프레임에서 'area' 열의 특정 값을 바꾸기
jeju_tour_alter['area'] = jeju_tour_alter['area'].replace({
    '성산일출봉(UNESCO 세계자연유산)': '성산일출봉',
    '제주동화마을': '송당리, 제주시',
    '한라산영실코스': '영실-윗세 코스',
    '우도(해양도립공원)': '우도, 연평리, 제주시',
    '송악산둘레길': '송악산, 서귀포시',
    '주상절리대(중문대포해안)': '제주중문대포해안 주상절리대 방문객센터',
    '애월한담해안산책로': 'Handam Coastal walk',
    '수월봉 지질트레일': '수월봉',
    '윗세오름(웃세오름)': '윗세오름',
    '신창풍차해안도로': 'Sinchang windmill, 한경해안로',
    '제주아트서커스': '동광로, 동광리, 서귀포시'


})

# 변경 결과 확인
print(jeju_tour_alter)


     Cnt                      area
9   7327                  송당리, 제주시
27  6745                       비자림
2   6381                  영실-윗세 코스
26  6000                     사려니숲길
21  5679                     성산일출봉
8   4417                  제주민속오일시장
13  3852                     용머리해안
12  3510              우도, 연평리, 제주시
22  3235                      섭지코지
28  2889                       만장굴
29  2854                    동문재래시장
19  2533                 송악산, 서귀포시
24  2521                      새별오름
5   2488      제주중문대포해안 주상절리대 방문객센터
0   2345                    협재해수욕장
18  2301                       쇠소깍
4   2134                     천지연폭포
1   2112                    함덕해수욕장
15  1830       Handam Coastal walk
17  1740                       수월봉
10  1738                      정방폭포
3   1613                     카멜리아힐
11  1495                      윗세오름
16  1483  Sinchang windmill, 한경해안로
7   1468            동광로, 동광리, 서귀포시
14  1440                   오설록티뮤지엄
25  1411                      산굼부리
23  1396            

In [19]:
jeju_tour_alter.to_csv('jeju_tour_alter.csv', index=False)

지명으로 위도와 경도를 불러오기 위해서 geopy 패키지 사용

In [20]:
pip install geopy

In [21]:
from geopy.geocoders import Nominatim #구글 네비게이션같은 패키지
import pandas as pd
import time

# Nominatim을 이용해 지오코더 객체 생성
geolocator = Nominatim(user_agent="jeju_locator")

# 빈 리스트를 만들어 위도와 경도 저장
latitude = []
longitude = []

# 각 관광지 이름을 통해 좌표를 찾고 리스트에 추가
for area in jeju_tour_alter['area']:
    try:
        # 지오코딩 요청: "지명, 제주도" 형식으로 검색
        location = geolocator.geocode(f"{area}, 제주도")
        if location:
            latitude.append(location.latitude)
            longitude.append(location.longitude)
        else:
            latitude.append(None)
            longitude.append(None)
        # 요청 간 딜레이 추가 (서비스 과부하 방지)
        time.sleep(1)
    except Exception as e:
        latitude.append(None)
        longitude.append(None)

# 위도와 경도를 데이터프레임에 새로운 열로 추가
jeju_tour_alter['Latitude'] = latitude
jeju_tour_alter['Longitude'] = longitude

# 결과 출력
print(jeju_tour_alter)


     Cnt                      area   Latitude   Longitude
9   7327                  송당리, 제주시  33.466280  126.775170
27  6745                       비자림  33.493177  126.810344
2   6381                  영실-윗세 코스  33.359115  126.501898
26  6000                     사려니숲길  33.423454  126.633267
21  5679                     성산일출봉  33.458889  126.940818
8   4417                  제주민속오일시장  33.493152  126.475312
13  3852                     용머리해안  33.233304  126.313067
12  3510              우도, 연평리, 제주시  33.507257  126.955002
22  3235                      섭지코지  33.436603  126.922083
28  2889                       만장굴  33.528254  126.770128
29  2854                    동문재래시장  33.512269  126.527630
19  2533                 송악산, 서귀포시  33.199081  126.289790
24  2521                      새별오름  33.366255  126.357688
5   2488      제주중문대포해안 주상절리대 방문객센터  33.237686  126.425957
0   2345                    협재해수욕장  33.394233  126.240817
18  2301                       쇠소깍  33.252594  126.623466
4   2134      

In [23]:
jeju_tour_loclat = jeju_tour_alter.copy()

In [24]:
jeju_tour_alter.to_csv('jeju_tour_loclat.csv', index=False)

#jeju_tour_loclat = pd.csv_read('')

folium 패키지로 관광지 반경 2km이내에 정류장 개수 count


In [25]:
!pip install folium

In [26]:
bus_jeju

정류장번호           정류장명         위도          경도       정보수집일  \
116191  JEB405000114         탑동푸른쉼터  33.516562  126.527116  2024-10-28   
116192  JEB405000115        탑동입구[서]  33.515313  126.524668  2024-10-28   
116193  JEB405000116        탑동입구[동]  33.515853  126.524825  2024-10-28   
116194  JEB405000117         남국사[동]  33.462152  126.549582  2024-10-28   
116195  JEB405000118         남국사[서]  33.460496  126.549790  2024-10-28   
...              ...            ...        ...         ...         ...   
130328  JEB405000301         천수동[동]  33.508467  126.541900  2024-10-28   
130329  JEB405000302         천수동[서]  33.509344  126.542568  2024-10-28   
130330  JEB405000303        국립제주박물관  33.512475  126.549973  2024-10-28   
130331  JEB405000304  고으니모르 국립제주박물관  33.512610  126.550965  2024-10-28   
130332  JEB405000305    거로 남쪽 입구[서]  33.504933  126.561300  2024-10-28   

        모바일단축번호  도시코드      도시명  관리도시명  
116191      NaN    39  제주특별자치도  제주BIS  
116192      NaN    39  제주특별자치도  제주BIS  
116193      NaN    39  제주특별자치도  제주BIS  
116194      NaN    39  제주특별자치도  제주BIS  
116195      NaN    39  제주특별자치도  제주BIS  
...         ...   ...      ...    ...  
130328      NaN    39  제주특별자치도  제주BIS  
130329      NaN    39  제주특별자치도  제주BIS  
130330      NaN    39  제주특별자치도  제주BIS  
130331      NaN    39  제주특별자치도  제주BIS  
130332      NaN    39  제주특별자치도  제주BIS  

[4195 rows x 9 columns]

In [44]:
bus_jeju.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4195 entries, 116191 to 130332
Data columns (total 9 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   정류장번호    4195 non-null   object 
 1   정류장명     4195 non-null   object 
 2   위도       4195 non-null   float64
 3   경도       4195 non-null   float64
 4   정보수집일    4195 non-null   object 
 5   모바일단축번호  0 non-null      float64
 6   도시코드     4195 non-null   int64  
 7   도시명      4195 non-null   object 
 8   관리도시명    4195 non-null   object 
dtypes: float64(3), int64(1), object(5)
memory usage: 456.8+ KB


In [27]:
jeju_tour_loclat

Cnt                      area   Latitude   Longitude
9   7327                  송당리, 제주시  33.466280  126.775170
27  6745                       비자림  33.493177  126.810344
2   6381                  영실-윗세 코스  33.359115  126.501898
26  6000                     사려니숲길  33.423454  126.633267
21  5679                     성산일출봉  33.458889  126.940818
8   4417                  제주민속오일시장  33.493152  126.475312
13  3852                     용머리해안  33.233304  126.313067
12  3510              우도, 연평리, 제주시  33.507257  126.955002
22  3235                      섭지코지  33.436603  126.922083
28  2889                       만장굴  33.528254  126.770128
29  2854                    동문재래시장  33.512269  126.527630
19  2533                 송악산, 서귀포시  33.199081  126.289790
24  2521                      새별오름  33.366255  126.357688
5   2488      제주중문대포해안 주상절리대 방문객센터  33.237686  126.425957
0   2345                    협재해수욕장  33.394233  126.240817
18  2301                       쇠소깍  33.252594  126.623466
4   2134                     천지연폭포  33.246927  126.554528
1   2112                    함덕해수욕장  33.543188  126.669879
15  1830       Handam Coastal walk  33.459474  126.310498
17  1740                       수월봉  33.294567  126.163241
10  1738                      정방폭포  33.244858  126.571545
3   1613                     카멜리아힐  33.289631  126.369889
11  1495                      윗세오름  33.360728  126.521027
16  1483  Sinchang windmill, 한경해안로  33.345464  126.171264
7   1468            동광로, 동광리, 서귀포시  33.316004  126.344597
14  1440                   오설록티뮤지엄  33.306055  126.289860
25  1411                      산굼부리  33.432078  126.694475
23  1396                 서귀포매일올레시장  33.250177  126.563570
6   1369                 제주절물자연휴양림  33.436729  126.627768
20  1366                    성이시돌목장  33.347146  126.326753

In [28]:
import folium
import pandas as pd
from IPython.display import IFrame #colab에서는 localhost 연결 불가


# 지도 생성 - 제주도 중심으로 설정
jeju_center = [33.4996, 126.5312]  # 제주도 중심 좌표
jeju_map = folium.Map(location=jeju_center, zoom_start=10)

# 각 관광지에 마커 추가
for index, row in jeju_tour_loclat.iterrows():
    folium.Marker(
        location=[row['Latitude'], row['Longitude']],
        popup=f"{row['area']} (방문자 수: {row['Cnt']})",
        icon=folium.Icon(color='blue', icon='info-sign')
    ).add_to(jeju_map)

# 지도 출력 (Colab에서는 HTML로 저장 후 iframe으로 표시)
jeju_map.save('jeju_tour_map.html')

# Google Colab에서 지도 보기 위해 iframe 사용
from IPython.display import IFrame

#IFrame('jeju_tour_map.html', width=700, height=500)
# Google Colab에서 저장된 HTML 파일을 보기 위해 iframe 사용
IFrame(src='./jeju_tour_map.html', width=700, height=500)

In [36]:
import folium
import pandas as pd

# 지도 생성 - 제주도 중심으로 설정
jeju_center = [33.4996, 126.5312]  # 제주도 중심 좌표
jeju_map = folium.Map(location=jeju_center, zoom_start=10)

# NaN이 아닌 위도와 경도를 가진 행만 필터링
jeju_tour_loclat_filtered = jeju_tour_loclat.dropna(subset=['Latitude', 'Longitude'])

# 각 관광지에 마커 추가
for index, row in jeju_tour_loclat_filtered.iterrows():
    folium.Marker(
        location=[row['Latitude'], row['Longitude']],
        popup=f"{row['area']} (방문자 수: {row['Cnt']})",
        icon=folium.Icon(color='orange', icon='star')
    ).add_to(jeju_map)

# Colab에서 Folium 지도 보기
jeju_map



In [43]:
import pandas as pd
from geopy.distance import geodesic

# bus_jeju 데이터프레임의 NaN 값 제거
bus_jeju = bus_jeju.dropna(subset=['위도', '경도'])

# 각 관광지에 대해 반경 2km 이내에 있는 버스 정류장의 수 계산
counts_within_2km = []  # 2km 내 반경 정류소 개수 리스트 초기화

for index, tour_row in jeju_tour_loclat_filtered.iterrows():
    jeju_tour_location = (tour_row['Latitude'], tour_row['Longitude'])

    # 위도와 경도가 올바른 범위 내에 있는지 확인
    if not (-90 <= jeju_tour_location[0] <= 90 and -180 <= jeju_tour_location[1] <= 180):
        counts_within_2km.append(0)
        continue

    count = 0
    for bus_index, bus_row in bus_jeju.iterrows():
        bus_location = (bus_row['위도'], bus_row['경도'])

        # 위도와 경도가 올바른 범위 내에 있는지 확인
        if not (-90 <= bus_location[0] <= 90 and -180 <= bus_location[1] <= 180):
            continue

        # 두 지점 사이의 거리 계산
        distance = geodesic(jeju_tour_location, bus_location).km
        if distance <= 2:
            count += 1
    counts_within_2km.append(count)

# 관광지 데이터프레임에 반경 2km 이내 버스 정류장 수 추가 (복사본에 추가하여 경고 방지)
jeju_tour_loclat_filtered = jeju_tour_loclat_filtered.copy()
jeju_tour_loclat_filtered['Bus_Count_within_2km'] = counts_within_2km

# 결과 확인
print(jeju_tour_loclat_filtered[['area', 'Bus_Count_within_2km']])


                        area  Bus_Count_within_2km
9                   송당리, 제주시                    31
27                       비자림                    13
2                   영실-윗세 코스                     0
26                     사려니숲길                     9
21                     성산일출봉                    13
8                   제주민속오일시장                   147
13                     용머리해안                    32
12              우도, 연평리, 제주시                     0
22                      섭지코지                    35
28                       만장굴                    11
29                    동문재래시장                   172
19                 송악산, 서귀포시                     4
24                      새별오름                     4
5       제주중문대포해안 주상절리대 방문객센터                    46
0                     협재해수욕장                    24
18                       쇠소깍                    54
4                      천지연폭포                   126
1                     함덕해수욕장                    49
15       Handam Coastal walk   

In [47]:
#반경 1km로 변경

import pandas as pd
from geopy.distance import geodesic

# bus_jeju 데이터프레임의 NaN 값 제거
bus_jeju = bus_jeju.dropna(subset=['위도', '경도'])

# 각 관광지에 대해 반경 1km 이내에 있는 버스 정류장의 수 계산
counts_within_1km = []  # 1km 내 반경 정류소 개수 리스트 초기화

for index, tour_row in jeju_tour_loclat_filtered.iterrows():
    jeju_tour_location = (tour_row['Latitude'], tour_row['Longitude'])

    # 위도와 경도가 올바른 범위 내에 있는지 확인
    if not (-90 <= jeju_tour_location[0] <= 90 and -180 <= jeju_tour_location[1] <= 180):
        counts_within_1km.append(0)
        continue

    count = 0
    for bus_index, bus_row in bus_jeju.iterrows():
        bus_location = (bus_row['위도'], bus_row['경도'])

        # 위도와 경도가 올바른 범위 내에 있는지 확인
        if not (-90 <= bus_location[0] <= 90 and -180 <= bus_location[1] <= 180):
            continue

        # 두 지점 사이의 거리 계산
        distance = geodesic(jeju_tour_location, bus_location).km
        if distance <= 1:
            count += 1
    counts_within_1km.append(count)

# 관광지 데이터프레임에 반경 1km 이내 버스 정류장 수 추가 (복사본에 추가하여 경고 방지)
jeju_tour_loclat_1km = jeju_tour_loclat_filtered.copy()
jeju_tour_loclat_1km['Bus_Count_within_1km'] = counts_within_1km

# 결과 확인
print(jeju_tour_loclat_1km[['area', 'Bus_Count_within_1km']])




                        area  Bus_Count_within_1km
9                   송당리, 제주시                    13
27                       비자림                     6
2                   영실-윗세 코스                     0
26                     사려니숲길                     4
21                     성산일출봉                     7
8                   제주민속오일시장                    38
13                     용머리해안                     9
12              우도, 연평리, 제주시                     0
22                      섭지코지                     6
28                       만장굴                     3
29                    동문재래시장                    57
19                 송악산, 서귀포시                     0
24                      새별오름                     1
5       제주중문대포해안 주상절리대 방문객센터                    12
0                     협재해수욕장                     8
18                       쇠소깍                     9
4                      천지연폭포                    42
1                     함덕해수욕장                    19
15       Handam Coastal walk   